storage/src/vespa/storage/storageserver/rpc/caching_rpc_target_resolver.cpp

```diff
@@ -1,82 +1,107 @@
+#include <cassert>
 
 #include <vespa/log/log.h>
 LOG_SETUP(".storage.caching_rpc_target_resolver");
 
 namespace storage::rpc {
 
-CachingRpcTargetResolver::CachingRpcTargetResolver(SharedRpcResources& rpc_resources)
-    : _rpc_resources(rpc_resources)
+CachingRpcTargetResolver::CachingRpcTargetResolver(const slobrok::api::IMirrorAPI& slobrok_mirror,
+                                                   const RpcTargetFactory& target_factory)
+    : _slobrok_mirror(slobrok_mirror),
+      _target_factory(target_factory),
+      _targets_rwmutex()
 {
 }
 
 CachingRpcTargetResolver::~CachingRpcTargetResolver() = default;
 
-namespace {
-
-vespalib::string address_to_slobrok_id(const api::StorageMessageAddress& address) {
+vespalib::string
+CachingRpcTargetResolver::address_to_slobrok_id(const api::StorageMessageAddress& address) {
     vespalib::asciistream as;
     as << "storage/cluster." << address.getCluster()
        << '/' << ((address.getNodeType() == lib::NodeType::STORAGE) ? "storage" : "distributor")
        << '/' << address.getIndex();
     return as.str();
 }
 
+std::shared_ptr<RpcTarget>
+CachingRpcTargetResolver::lookup_target(const vespalib::string& slobrok_id, uint32_t curr_slobrok_gen) {
+    std::shared_lock lock(_targets_rwmutex);
+    auto itr = _targets.find(slobrok_id);
+    if ((itr != _targets.end())
+        && itr->second->_target->is_valid()
+        && (itr->second->_slobrok_gen == curr_slobrok_gen)) {
+        return itr->second;
+    }
+    return {};
+}
+
+std::shared_ptr<RpcTarget>
+CachingRpcTargetResolver::consider_update_target(const vespalib::string& slobrok_id,
+                                                 const vespalib::string& connection_spec,
+                                                 uint32_t curr_slobrok_gen,
+                                                 const UniqueLock& targets_lock) {
+    (void) targets_lock;
+    // If address has the same spec as the existing target, just reuse it.
+    auto itr = _targets.find(slobrok_id);
+    if ((itr != _targets.end())
+        && (itr->second->_target->is_valid())
+        && (itr->second->_spec == connection_spec))
+    {
+        LOG(info, "Updating existing mapping '%s' -> '%s' (gen %u) to gen %u",
+            slobrok_id.c_str(), connection_spec.c_str(), itr->second->_slobrok_gen, curr_slobrok_gen);
+        itr->second->_slobrok_gen = curr_slobrok_gen;
+        return itr->second;
+    }
+    return {};
+}
+
+std::shared_ptr<RpcTarget>
+CachingRpcTargetResolver::insert_new_target_mapping(const vespalib::string& slobrok_id,
+                                                    const vespalib::string& connection_spec,
+                                                    uint32_t curr_slobrok_gen,
+                                                    const UniqueLock& targets_lock) {
+    (void) targets_lock;
+    auto target = _target_factory.make_target(connection_spec, curr_slobrok_gen); // TODO expensive inside lock?
```

Comment:
Should we consider having the factory return `shared_ptr` instead of `unique_ptr`? The latter is more obvious that returned objects are unique, the former avoids an extra control block allocation due to `make_shared`.

Reply:
Discussed offline. We stick with `unique_ptr` as that better describes intent.